modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/JSResolveModulesServlet.java

```diff
@@ -54,12 +57,27 @@
 )
 public class JSResolveModulesServlet extends HttpServlet {
 
+	public JSResolveModulesServlet() {
+		updateEtag();
+	}
+
 	@Override
 	protected void service(
 			HttpServletRequest httpServletRequest,
 			HttpServletResponse httpServletResponse)
 		throws IOException {
 
+		if (_etag.equals(
```

Comment:
Why do we need to implement this ourselves rather than relying on the `ETagFilter` that's already processing requests?

I think you tried to explain offline, but I'm still not sure I understand :)

Reply:
Because `ETagFilter` does a checksum with the content and simply adds the result as an HTTP header, but that doesn't fit in our scenario.

We don't need a strong hash-based ETag, just a weak (thus the `W/` prefix) one which simply designates the _revision number_ of the current state of the `NPMRegistry`. 

All contents returned by `/js_resolve_modules` will have the same `ETag` until the `NPMRegistry` state changes. When that happens, a new UUID will be generated so that all cached resolutions are invalidated and everybody receives new copies again when they hit `/js_resolve_modules`.

This way we don't need to:

1) compute any hash (saving a lot of CPU when the browser and server ETags match)
2) define any revision number management algorithm (because we use random UUID and that guarantees that they will be unique in time)
3) maintain any table in the server to match ETags to URLs (in case we wanted to optimize 1)

Also, the configuration and implementation of the `ETagFilter` and `CacheFilter` relies on companyIds, groupIds, and so on, which doesn't apply in this case so it seems much safer to control the caching in this same servlet rather than relying on external filters that can be tweaked or changed without us noticing it. However, note that this has not been the main reason to choose this strategy, but the reasons stated above, which can be summarized as "simplicity".

Reply:
More information about ETag generation can be found in the [Wikipedia](https://en.wikipedia.org/wiki/HTTP_ETag#ETag_generation) which explains the subject quite nicely.

Reply:
Ok, let's give this a try then!